README
= Back Door, execute Ruby code directly inside Radiant templates

==Intro

Back door is a Radiant extension that allows executing Ruby code directly
inside Radiant templates. It provides the following main tags: <r:ruby>, <r:erb>,
<r:if>, <r:else>, <r:unless>, <r:tag> and <r:erb_tag>.

The objective of this extension is to avoid writing Radiant extensions for easy
tasks.  If you just need some Ruby/ERB code inside templates, this extension
is for you.  If you want to implement Radius tags, modify the admin interface or
define some other complex behavior you should revert to writing a complete Radiant
extension.

Download::          git://github.com/aitorgr/radiant-backdoor.git

Project home::      http://github.com/aitorgr/radiant-backdoor
Current version::   0.4.2
Author::            aitor.name[http://aitor.name]

Please discuss, report bugs and comments to the Radiant user's mailing-list[http://radiantcms.org/mailing-list]
or contact the author directly at me(at)aitor.name.

Thanks to RubyForge for hosting this project.

== Usage

This extension adds the following new tags:

=== <r:ruby>

Executes the content of the tag body as Ruby code and renders the returned value.

*Usage:*

   <r:ruby> [ruby code] </r:ruby> 

*Example:*

   With this fine extension you have access to all this information: <r:ruby> self.inspect </r:ruby> 
   This machine identifies itself as <r:ruby> `uname -a` </r:ruby> 
  
=== <r:erb>

Executes the content of the tag body as ERB code and renders the returned value.

*Usage:*

   <r:erb> [ERB template] </r:erb> 

*Example:*

  
  <r:erb>
    <table>
      <tr>
        <th>name</th> <th>password</th> <th>uid</th> <th>gid</th> <th>class</th> <th>home_dir</th> <th>shell</th>
      </tr>
      <%  File.open( "/etc/passwd") do |io|
            while line = io.gets
              next if line =~ /^\s*#/
              fields = line.split( ":")
      %>
              <tr>
               <% fields.each do |field| %>
                <td> <%= field %> </td>
              <% end %>
              </tr>
      <%    end
          end
      %>
    </table>
  </r:erb>
  

*Caveats:*

When using ruby looping constructs inside a ERB template, the Radius tags present in the template get expanded
only once, and then "repeated" multiple times by the Ruby code.  For example, the following code:

  <r:erb>
    <% 5.times do %>
      <%= <r:cycle values="1, 2, 3, 4, 5"/> %>
    <% end %>
  </r:erb>

returns "11111", though intuitively it should return "12345" (ignoring whitespace).  This is because the <r:cycle>
tag gets expanded only once.

Nearly always this is irrelevant, since Radiant tags are all side effect free, that is, they always return the same value
when called multiple times in the same context.  Notable exceptions are the <r:cycle> and <r:random> tags.

For the case of tags with side effects, use the <r:expand> tag as documented in the description for that tag.
  
=== <r:expand>

When used inside an "r:erb" tag, it expands the tag named in the "tag" attribute with the given attributes.

This tag avoids the "tag only expands once inside ruby loops" problem described in the "r:erb" tag.  It takes a required
"tag" attribute that identifies an existing tag, and expands that tag passing it the rest of attributes.

NOTE that this tag must be used inside an ERB <%= %> construct.

*Usage:"

   <r:expand tag="tag-name" [ tag-name-attribute="value"]* /> 

*Example:"

  <r:erb>
    <% <r:cycle values="1, 2, 3, 4, 5" reset="true"/> %>
    <% 5.times do %>
      <%=
      <r:expand tag="cycle" values="1, 2, 3, 4, 5"/>
      %>
    <% end %>
  </r:erb>

renders "23451" (ignoring whitespace)
  
=== <r:if>

Renders the tag body if the given Ruby expression evaluates to true. If not it renders an inner "r:else" tag, if present. Note that there are some caveats regarding the "r:else" tag, look the description of if for more information.

*Usage:*

   <r:if cond="[ruby expression]"> [HTML content] </r:if> 
   <r:if cond="[ruby expression]"> [HTML content] <r:else> [HTML content] </r:else> </r:if> 

*Example:*

  
  <r:if cond="request.env[ 'HTTP_USER_AGENT'] =~ /MSIE/">
    <!-- Internet Explorer needs some ugly hacks to render PNG images with transparency -->
    <div id="logo" style="filter:progid:DXImageTransform.Microsoft.AlphaImageLoader(src=\'/images/logo.png\',sizingMethod=\'scale\');"></div>
    <r:else>
      <!-- the rest of browsers just make it rigth -->
      <img src="/images/logo.png" id="logo" alt="Logo"/>
    </r:else>
  </r:if>
  

*Caveats:*

There are two caveats with the "r:if/unless" and "r:else" tags (used "r:if" as example, but equally applicable to "r:unless"):

* to allow the expansion of an "r:else" part inside a "r:if" tag, the containing "r:if" tag must be
  expanded even if the condition of the "r:if" tag evaluates to false.  In this case, the body of the
  "r:if" tag is ignored.  This is not a problem unless the body of the "r:if" tag has "side effects":

    
    <r:ruby>
      @counter = 0
      ""
    </r:ruby>
    <r:if cond="false">
      This text is evaluated and ignored, but the following "ruby" tag has a side effect that affects the "else" tag
      <r:ruby> @counter += 1 </r:ruby>
      <r:else> <r:ruby> @counter </r:ruby> </r:else>
    </r:if>
    

* the rendered value of the "r:if" tag contains also any data after an "r:else" tag:

    
    <r:if cond="true">
      Hello
      <r:else> Hello Radiant! </r:else>
      world!
    </r:if>
    

  evaluates to "Hello world!" (ignoring whitespace)
  
=== <r:unless>

The inverse of the "r:if" tag.  Refer to the description of the "r:if" tag for complete documentation.
  
=== <r:else>

Specifies the "else" part for "r:if" and "r:unless".  Refer to the description of the "r:if" tag for complete documentation.
  
=== <r:tag>

Defines a new tag, which can subsequently be used as a normal Radius tag.  It can be seen as a programmable and parametrized snippet.

The new tag receives the tag context with the standard name "tag" (see the example below).  Note that this tag must be rendered by
Radiant before the tag it defines is used.  This typically means that this tag must be used early in a page layout.

*Usage:*

 

  <!-- define a new tag -->
  <r:tag name="tag-name">
    [Ruby code]
  </r:tag>

  <!-- use the new defined tag -->
  <r:tag-name [params]>
    [content]
  </r:tag-name>



*Example:*



  <!-- define a new tag -->
  <r:tag name="originating_ip">
    @request.remote_ip
  </r:tag>

  <!-- use it -->
  Article posted from IP <r:originating_ip />


  
=== <r:erb_tag>

Same functionality as the "tag" tag but the body of the tag is interpreted as ERB code.  Handy for heavy parametrized templating.

*Usage:*

 

  <!-- define a new tag -->
  <r:tag name="tag-name">
    [Ruby code]
  </r:tag>

  <!-- use the new defined tag -->
  <r:tag-name [params]>
    [content]
  </r:tag-name>



*Example:*



  <!-- define a new tag -->
  <r:erb_tag name="article">
    <div class="<%= tag.attr[ "class"] || "article" %>">
      <div class="article-title"> <%= tag.attr[ "title"] %> </div>
      <div class="article-body"> <%= tag.expand %> </div>
      <div class="article-footer"> Posted in <page/> </div>
    </div>
  </r:erb>

  <!-- use it -->
  <article title="New BackDoor release">
    Blah, blah, blah
  </article>


  


== Attribute expansion

Back Door allows evaluating tag's attributes as Ruby code.  If a tag's attribute starts with the "=" symbol, the rest of
the attribute is evaluated, and the attribute's value is replaced with the evaluated value.  This works for both standard
Radiant (Radius) tags and Back Door's own tags.

As an example, consider a side-bar where we want to show first and second level pages.  But for the "news" page, we want to
limit the number of children to 3 and add an extra link to see all news.  Without Back Door this could be implemented as:

  <r:children:each>
      <h2 class="menu-heading">
          <r:link>
              <r:title />
          </r:link>
      </h2>
      <r:if_url matches="/news/">
          <r:children:each limit="3" order="desc" by="published_at">
              <h3 class="menu-subheading">
                  <r:link>
                      <r:title />
                  </r:link>
              </h3>
          </r:children:each>
          <h3 class="menu-subheading">
              <r:link> More news... </r:link>
          </h3>
      </r:if_url>
      <r:unless_url matches="/news/">
          <r:children:each order="desc" by="published_at">
              <h3 class="menu-subheading">
                  <r:link>
                      <r:title />
                  </r:link>
              </h3>
          </r:children:each>
      </r:unless_url>
  </r:children:each>

Clearly there is too much duplication.  With Back Door this can be rewritten as:

  <r:children:each>
      <h2 class="menu-heading">
          <r:link>
              <r:title />
          </r:link>
      </h2>
      <r:ruby> @news = tag.locals.page.url =~ /news/; "" </r:ruby>
      <r:children:each limit="=(@news ? 3: 1000)" order="desc" by="published_at">
          <h3 class="menu-subheading">
              <r:link>
                  <r:title />
              </r:link>
          </h3>
      </r:children:each>
      <r:if cond="@news">
          <h3 class="menu-subheading">
              <r:link> More news... </r:link>
          </h3>
      </r:if>
  </r:children:each>

Note how the "limit" attribute of the <r:children:each> tag uses the "=" symbol to make it dynamic.

== Security

None to date.  Arbitrary Ruby code can be executed with this extension.  If you don't trust the people
writing the pages in a Radiant site armed with this extension, simply don't use it.  See the TODO[link:files/TODO.html] file
for more information.

== License

The Back Door extension is available under a MIT style license.

Copyright (c) 2007 Aitor Garay-Romero <me(at)aitor.name>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGELOG
= Change Log

=== 0.4.3
* Use "=" character for attribute expansion instead of "#" to avoid conflicts with HTML anchors.  Thanks to Martin Grunskis <one.three(at)gmail.com> for the bug report and the discussion that lead to the problem.

=== 0.4.2
* Check that we don't activate the extension twice, this happens in development environment

=== 0.4.1
* Automatically convert to string the result of evaluating tag attributes, so the user does not have to take care of this

=== 0.4.0
* Allow expansion of tag attributes by prepending them with the "#" symbol, read the "Attribute expansion" section of the README for more information

=== 0.3.0
* Implemented a <r:else> tag for both <r:if> and <r:unless> tags.  Now it's possible to do: 

  	<r:if cond="...">
  		This is the if part
		<r:else>
			And this the else part
		</r:else>
	</r:if>

* Implemented a <r:tag> and <r:erb_tag> tags which allow to define new tags easily:

	<!-- define a new tag -->
	<r:erb_tag name="article">
		<div class="<%= tag.attr[ "class"] || "article" %>">
			<div class="article-title"> <%= tag.attr[ "title"] %> </div>
			<div class="article-body"> <%= tag.expand %> </div>
			<div class="article-footer"> Posted in <page/> </div>
		</div>
	</r:erb>

	<!-- use it -->
	<article title="New BackDoor release">
		Blah, blah, blah
	</article>

* Added a <r:erb:expand> tag that solves a problem where tags inside ERB templates expand only once when used inside ruby loops 

=== 0.2.0
* Implemented the <r:unless> tag.  Thanks to Jamie Wilkinson <jamie-list(at)tramchase.com> for providing the patch.
* Check that tests are executed inside a Radiant application.
* Minor documentation clean up.

=== 0.1.0
* First release.  Implemented tags <r:ruby>, <r:erb> and <r:if>.


test/test_helper.rb
# Load the the environment
unless defined? RADIANT_ROOT
  ENV["RAILS_ENV"] = "test"
  boot_file = "#{File.expand_path(File.dirname(__FILE__) + "/../../../../")}/config/boot"
  raise "Tests for this extension must be executed when it's properly installed in a radiant app" unless File.exist?( boot_file + ".rb")
  require boot_file
end
require "#{RADIANT_ROOT}/test/test_helper"

class Test::Unit::TestCase
  
  # Include a helper to make testing Radius tags easier
  test_helper :extension_tags
  
  # Add the fixture directory to the fixture path
  self.fixture_path << File.dirname(__FILE__) + "/fixtures"
  
  # Add more helper methods to be used by all extension tests here...
  
end


test/unit/back_door_tags_test.rb
require File.dirname(__FILE__) + '/../test_helper'

# *noTE* be careful with spaces into the template strings

class BackDoorTagsTest < Test::Unit::TestCase

  fixtures    :pages
  test_helper :render

  def setup
    @page = pages( :radius)
  end

  def test_ruby_tag
    assert_renders "hello", %q{<r:ruby> "hello" </r:ruby>} 
    assert_renders "0", %q{<r:ruby> 0 </r:ruby>}
    assert_renders "0123456789", %q{<r:ruby> (0..9).inject( "") { |acc,n| acc + n.to_s } </r:ruby>}

    # test that backdoor tags are executed in the same object context
    assert_render_match /\s+(\S+)\s+(\1)/, %q{
      <r:ruby> self.to_s </r:ruby>
      <r:ruby> self.to_s </r:ruby>
    }
  end

  def test_erb_tag
    assert_renders "hello", %q{<r:erb><%= "hello" %></r:erb>}
    assert_renders "0", %q{<r:erb><% n = 0 %><%= n %></r:erb>}
    assert_renders "0123456789", %q{<r:erb>0<%= (1..8).inject( "") { |acc,n| acc + n.to_s }%>9</r:erb>}

    # this test should use a RE of /\s*2\s*3\s*4\s*5\s*1/, see the doc for the <r:erb> tag
    assert_render_match /\s*2\s*2\s*2\s*2\s*2/, %q{
      <r:erb>
        <% <r:cycle values="1, 2, 3, 4, 5" reset="true"/> %>
        <% 5.times do %>
          <%= <r:cycle values="1, 2, 3, 4, 5"/> %>
        <% end %>
      </r:erb>
    }

    assert_render_match /\s*2\s*3\s*4\s*5\s*1/, %q{
      <r:erb>
        <% <r:cycle values="1, 2, 3, 4, 5" reset="true"/> %>
        <% 5.times do %>
          <%=
          <r:expand tag="cycle" values="1, 2, 3, 4, 5"/>
          %>
        <% end %>
      </r:erb>
    }
  end

  def test_if_tag

    # if
    assert_renders "true", %q{<r:if cond="true">true</r:if>}
    assert_renders "", %q{<r:if cond="false">true</r:if>}
    assert_raise BackDoorTags::TagError do
      assert_renders "", %q{<r:if con="false">true</r:if>}
    end

    # else
    assert_renders "true", %q{<r:if cond="true">true<r:else>false</r:else></r:if>}
    assert_renders "truetrue", %q{<r:if cond="true">true<r:else>false</r:else>true</r:if>}
    assert_renders "false", %q{<r:if cond="false">true<r:else>false</r:else></r:if>}
    assert_renders "false", %q{<r:if cond="false">true<r:else>false</r:else>true</r:if>}

    # side effects
    assert_render_match /\s*1\s*/, %q{
      <r:ruby>
        @counter = 0
        ""
      </r:ruby>
      <r:if cond="false">
        This text is evaluated and igored, but the following "ruby" tag has a side effect that affects the "else" tag
        <r:ruby> @counter += 1 </r:ruby>
        <r:else> <r:ruby> @counter </r:ruby> </r:else>
      </r:if>
    }
  end

  def test_unless_tag

    # unless
    assert_renders "false", %q{<r:unless cond="false">false</r:unless>}
    assert_renders "", %q{<r:unless cond="true">true</r:unless>}
    assert_raise BackDoorTags::TagError do
      assert_renders "", %q{<r:unless con="true">false</r:unless>}
    end

    # else
    assert_renders "false", %q{<r:unless cond="true">true<r:else>false</r:else></r:unless>}
    assert_renders "false", %q{<r:unless cond="true">true<r:else>false</r:else>true</r:unless>}
    assert_renders "true", %q{<r:unless cond="false">true<r:else>false</r:else></r:unless>}
    assert_renders "truetrue", %q{<r:unless cond="false">true<r:else>false</r:else>true</r:unless>}
  end

  def test_tag_definition

    assert_render_match /01234567/, %q{
      <r:tag name="test_tag">
        raise TagError.new( "'test_tag' tag must contain a 'up_to' attribute.") unless tag.attr.has_key?( "up_to")
        up_to = tag.attr[ "up_to"]
        (0..up_to.to_i).inject( "") { |acc,n| acc + n.to_s }
      </r:tag>
      <r:test_tag up_to="7" />
    }

    assert_render_match /(\s*Hello World!\s*){3}/, %q{
      <r:erb_tag name="test_tag">
        <% tag.attr[ "times"].to_i.times do %>
          Hello <%= [ "W", "o", "r", "l", "d", "!"].join %>
        <% end %>
      </r:erb_tag>
      <r:test_tag times="3" />
    }
  end

  def test_attribute_expansion
    assert_render_match /.*?(#2).*?(id-4)/, %q{
      <r:link anchor="=(1+1)" id='="id-#{(2+2)}"' />
    }
    assert_render_match /\s+(\S+)\s+(\1)/, %q{
      <r:ruby> self.to_s </r:ruby>
      <r:ruby attr="=self.to_s">
        tag.attr[ "attr"]
      </r:ruby>
    }
    # check that old expansion prefix does not work now
    assert_render_match /#\(1\+1\)/, %q{
      <r:link anchor="#(1+1)" />
    }
  end

  def test_attribute_expansion_in_a_loop
    assert_render_match /testtest/, %q{
      <r:ruby> @counter = 0; "" </r:ruby>
      <r:children:each>
        <r:link id="=@counter" />
        <r:ruby> @counter +=1; "" </r:ruby>
      </r:children:each>
    }
  end

end

=begin
      <r:erb>
        <% @counter = 1 %>
        -- <r:link id='=@counter' /> -- 
        <% 5.times do %>
          <%= @counter %>
          <%= %q{<r:link id='=0' />} %>
          <%# @counter %>
          <% @counter += 1 %>
        <% end %>
      </r:erb>
=end


test/functional/back_door_extension_test.rb
require File.dirname(__FILE__) + '/../test_helper'

class BackDoorExtensionTest < Test::Unit::TestCase
  
  def test_this_extension
    assert true
  end
  
=begin
  def test_initialization
    assert_equal File.join(File.expand_path(RAILS_ROOT), 'vendor', 'extensions', 'back_door'), BackDoorExtension.root
    assert_equal 'Back Door', BackDoorExtension.extension_name
  end
=end
  
end


app/models/back_door_tags.rb
module BackDoorTags
  include Radiant::Taggable

  class TagError < StandardError; end

  # don't remove the comment below, it's used by README.erb
  # ruby
  desc %{
    Executes the content of the tag body as Ruby code and renders the returned value.

    *Usage:*

      <pre><code> <r:ruby> [ruby code] </r:ruby> </code></pre>

    *Example:*

      <pre><code> With this fine extension you have access to all this information: <r:ruby> self.inspect </r:ruby> </code></pre>
      <pre><code> This machine identifies itself as <r:ruby> `uname -a` </r:ruby> </code></pre>
  }
  tag "ruby" do |tag|
    eval( tag.expand)
  end

  # don't remove the comment below, it's used by README.erb
  # erb
  desc %{
    Executes the content of the tag body as ERB code and renders the returned value.

    *Usage:*

      <pre><code> <r:erb> [ERB template] </r:erb> </code></pre>

    *Example:*

      <pre><code>
      <r:erb>
        <table>
          <tr>
            <th>name</th> <th>password</th> <th>uid</th> <th>gid</th> <th>class</th> <th>home_dir</th> <th>shell</th>
          </tr>
          <%  File.open( "/etc/passwd") do |io|
                while line = io.gets
                  next if line =~ /^\s*#/
                  fields = line.split( ":")
          %>
                  <tr>
                   <% fields.each do |field| %>
                    <td> <%= field %> </td>
                  <% end %>
                  </tr>
          <%    end
              end
          %>
        </table>
      </r:erb>
      </code></pre>

    *Caveats:*

    When using ruby looping constructs inside a ERB template, the Radius tags present in the template get expanded
    only once, and then "repeated" multiple times by the Ruby code.  For example, the following code:

      <r:erb>
        <% 5.times do %>
          <%= <r:cycle values="1, 2, 3, 4, 5"/> %>
        <% end %>
      </r:erb>

    returns "11111", though intuitively it should return "12345" (ignoring whitespace).  This is because the <r:cycle>
    tag gets expanded only once.

    Nearly always this is irrelevant, since Radiant tags are all side effect free, that is, they always return the same value
    when called multiple times in the same context.  Notable exceptions are the <r:cycle> and <r:random> tags.

    For the case of tags with side effects, use the <r:expand> tag as documented in the description for that tag.
  }
  tag "erb" do |tag|
    ERB.new( tag.expand).result( binding)
  end

  # don't remove the comment below, it's used by README.erb
  # expand
  desc %{
    When used inside an "r:erb" tag, it expands the tag named in the "tag" attribute with the given attributes.

    This tag avoids the "tag only expands once inside ruby loops" problem described in the "r:erb" tag.  It takes a required
    "tag" attribute that identifies an existing tag, and expands that tag passing it the rest of attributes.

    NOTE that this tag must be used inside an ERB <%= %> construct.

    *Usage:"

      <pre><code> <r:expand tag="tag-name" [ tag-name-attribute="value"]* /> </code></pre>

    *Example:"

      <r:erb>
        <% <r:cycle values="1, 2, 3, 4, 5" reset="true"/> %>
        <% 5.times do %>
          <%=
          <r:expand tag="cycle" values="1, 2, 3, 4, 5"/>
          %>
        <% end %>
      </r:erb>

    renders "23451" (ignoring whitespace)
  }
  tag "erb:expand" do |tag|
    raise TagError.new( "'expand' tag must contain a 'tag' attribute.") unless tag.attr.has_key?( "tag")
    tag_name = tag.attr.delete( "tag")
    "tag.render( \"#{tag_name}\", #{tag.attr.inspect})"
  end

  # don't remove the comment below, it's used by README.erb
  # if
  desc %{
    Renders the tag body if the given Ruby expression evaluates to true. If not it renders an inner "r:else" tag, if present. Note that there are some caveats regarding the "r:else" tag, look the description of if for more information.
    
    *Usage:*

      <pre><code> <r:if cond="[ruby expression]"> [HTML content] </r:if> </code></pre>
      <pre><code> <r:if cond="[ruby expression]"> [HTML content] <r:else> [HTML content] </r:else> </r:if> </code></pre>

    *Example:*

      <pre><code>
      <r:if cond="request.env[ 'HTTP_USER_AGENT'] =~ /MSIE/">
        <!-- Internet Explorer needs some ugly hacks to render PNG images with transparency -->
        <div id="logo" style="filter:progid:DXImageTransform.Microsoft.AlphaImageLoader(src=\'/images/logo.png\',sizingMethod=\'scale\');"></div>
        <r:else>
          <!-- the rest of browsers just make it rigth -->
          <img src="/images/logo.png" id="logo" alt="Logo"/>
        </r:else>
      </r:if>
      </code></pre>

    *Caveats:*

    There are two caveats with the "r:if/unless" and "r:else" tags (used "r:if" as example, but equally applicable to "r:unless"):

    * to allow the expansion of an "r:else" part inside a "r:if" tag, the containing "r:if" tag must be
      expanded even if the condition of the "r:if" tag evaluates to false.  In this case, the body of the
      "r:if" tag is ignored.  This is not a problem unless the body of the "r:if" tag has "side effects":

        <pre><code>
        <r:ruby>
          @counter = 0
          ""
        </r:ruby>
        <r:if cond="false">
          This text is evaluated and ignored, but the following "ruby" tag has a side effect that affects the "else" tag
          <r:ruby> @counter += 1 </r:ruby>
          <r:else> <r:ruby> @counter </r:ruby> </r:else>
        </r:if>
        </code></pre>

    * the rendered value of the "r:if" tag contains also any data after an "r:else" tag:

        <pre><code>
        <r:if cond="true">
          Hello
          <r:else> Hello Radiant! </r:else>
          world!
        </r:if>
        </code></pre>

      evaluates to "Hello world!" (ignoring whitespace)
  }
  tag "if" do |tag|
    raise TagError.new( "'if' tag must contain a 'cond' attribute.") unless tag.attr.has_key?( "cond")
    cond = eval( tag.attr[ "cond"])
    tag.globals.else_cond = !cond       # we pass to the <r:else> tag the flag that specifies if it should evaluate or not
    tag.globals.else_result = ""        # an <r:else> tag inside this <r:if> tag can modify this
    if_result = tag.expand              # we expand the tag in any case, this is because the <r:else> tag inside (if any) must be evaluated
    return cond ? if_result: tag.globals.else_result
  end

  # don't remove the comment below, it's used by README.erb
  # unless
  desc %{
    The inverse of the "r:if" tag.  Refer to the description of the "r:if" tag for complete documentation.
  }
  tag "unless" do |tag|
    raise TagError.new( "'unless' tag must contain a 'cond' attribute.") unless tag.attr.has_key?( "cond")
    cond = eval( tag.attr[ "cond"])
    tag.globals.else_cond = cond        # we pass to the <r:else> tag the flag that specifies if it should evaluate or not
    tag.globals.else_result = ""        # an <r:else> tag inside this <r:else> tag can modify this
    if_result = tag.expand              # we expand the tag in any case, this is because the <r:else> tag inside (if any) must be evaluated
    return cond ? tag.globals.else_result: if_result
  end

  # don't remove the comment below, it's used by README.erb
  # else
  desc %{
    Specifies the "else" part for "r:if" and "r:unless".  Refer to the description of the "r:if" tag for complete documentation.
  }
  [ "if:else", "unless:else"].each do |tag_id|
    tag tag_id do |tag|
      if tag.globals.else_cond
        tag.globals.else_result = tag.expand
      end
      ""
    end
  end

  # don't remove the comment below, it's used by README.erb
  # tag
  desc %{
    Defines a new tag, which can subsequently be used as a normal Radius tag.  It can be seen as a programmable and parametrized snippet.

    The new tag receives the tag context with the standard name "tag" (see the example below).  Note that this tag must be rendered by
    Radiant before the tag it defines is used.  This typically means that this tag must be used early in a page layout.
    
    *Usage:*

    <pre><code> 

      <!-- define a new tag -->
      <r:tag name="tag-name">
        [Ruby code]
      </r:tag>

      <!-- use the new defined tag -->
      <r:tag-name [params]>
        [content]
      </r:tag-name>

    </code></pre>

    *Example:*

    <pre><code>

      <!-- define a new tag -->
      <r:tag name="originating_ip">
        @request.remote_ip
      </r:tag>

      <!-- use it -->
      Article posted from IP <r:originating_ip />

    </code></pre>
  }
  tag "tag" do |external_tag|
    raise TagError.new( "'tag' tag must contain a 'name' attribute.") unless external_tag.attr.has_key?( "name")
    code = external_tag.expand
    external_tag.context.define_tag( external_tag.attr[ "name"]) do |tag|
      eval( code)
    end
  end

  # don't remove the comment below, it's used by README.erb
  # erb_tag
  desc %{
    Same functionality as the "tag" tag but the body of the tag is interpreted as ERB code.  Handy for heavy parametrized templating.

    *Usage:*

    <pre><code> 

      <!-- define a new tag -->
      <r:tag name="tag-name">
        [Ruby code]
      </r:tag>

      <!-- use the new defined tag -->
      <r:tag-name [params]>
        [content]
      </r:tag-name>

    </code></pre>

    *Example:*

    <pre><code>

      <!-- define a new tag -->
      <r:erb_tag name="article">
        <div class="<%= tag.attr[ "class"] || "article" %>">
          <div class="article-title"> <%= tag.attr[ "title"] %> </div>
          <div class="article-body"> <%= tag.expand %> </div>
          <div class="article-footer"> Posted in <page/> </div>
        </div>
      </r:erb>

      <!-- use it -->
      <article title="New BackDoor release">
        Blah, blah, blah
      </article>

    </code></pre>
  }
  tag "erb_tag" do |external_tag|
    raise TagError.new( "'tag' tag must contain a 'name' attribute.") unless external_tag.attr.has_key?( "name")
    code = external_tag.expand
    external_tag.context.define_tag( external_tag.attr[ "name"]) do |tag|
      ERB.new( code).result( binding)
    end
  end
end
